lang/c++/impl/avrogencpp.cc

```diff
@@ -915,3 +924,34 @@ int main(int argc, char **argv) {
         return 1;
     }
 }
+
+UnionCodeTracker::UnionCodeTracker(const std::string &schemaFile) : schemaFile_(schemaFile) {
+}
+
+std::optional<std::string> UnionCodeTracker::getExistingUnionName(const std::vector<std::string> &unionBranches) const {
+    if (const auto it = unionBranchNameMapping_.find(unionBranches); it != unionBranchNameMapping_.end()) {
+        return it->second;
+    }
+    return std::nullopt;
+}
+
+std::string UnionCodeTracker::generateNewUnionName(const std::vector<std::string> &unionBranches) {
+    string s = schemaFile_;
+    string::size_type n = s.find_last_of("/\\");
+    if (n != string::npos) {
+        s = s.substr(n);
+    }
+    makeCanonical(s, false);
+
+    std::string result = s + "_Union__" + boost::lexical_cast<string>(unionNumber_++) + "__";
+    unionBranchNameMapping_.emplace(unionBranches, result);
+    return result;
+}
+
+bool UnionCodeTracker::unionTraitsAlreadyGenerated(const std::string &unionClassName) const {
+    return generatedUnionTraits_.find(unionClassName) != generatedUnionTraits_.end();
+}
+
+void UnionCodeTracker::setTraitsGenerated(const std::string &unionName) {
```

Comment:
For consistency I'd suggest to use the same name for the parameter `unionName` -> `unionClassName`.

Reply:
Good Point, i renamed it